pkg/configurator/crd_client.go

```diff
@@ -0,0 +1,183 @@
+	"github.com/openservicemesh/osm/pkg/gen/client/config/clientset/versioned"
+	informers "github.com/openservicemesh/osm/pkg/gen/client/config/informers/externalversions"
+	k8s "github.com/openservicemesh/osm/pkg/kubernetes"
+	"github.com/openservicemesh/osm/pkg/kubernetes/events"
+)
+
+const (
+	meshConfigInformerName = "MeshConfig"
+	meshConfigProviderName = "OSM"
+)
+
+func newConfiguratorWithCRDClient(meshConfigClientSet versioned.Interface, stop <-chan struct{}, osmNamespace string) *CRDClient {
+	informerFactory := informers.NewSharedInformerFactoryWithOptions(
+		meshConfigClientSet,
+		k8s.DefaultKubeEventResyncInterval,
+		informers.WithNamespace(osmNamespace),
+	)
+	informer := informerFactory.Config().V1alpha1().MeshConfigs().Informer()
+	crdClient := CRDClient{
+		informer:     informer,
+		cache:        informer.GetStore(),
+		cacheSynced:  make(chan interface{}),
+		osmNamespace: osmNamespace,
+	}
+
+	// configure listener
+	eventTypes := k8s.EventTypes{
+		Add:    announcements.MeshConfigAdded,
+		Update: announcements.MeshConfigUpdated,
+		Delete: announcements.MeshConfigDeleted,
+	}
+	informer.AddEventHandler(k8s.GetKubernetesEventHandlers(meshConfigInformerName, meshConfigProviderName, nil, eventTypes))
+
+	// start listener
+	crdClient.startMeshConfigListener(stop)
+
+	// start informer
+	go informer.Run(stop)
+
+	return &crdClient
+}
+
+// Listens to ConfigMap events and notifies dispatcher to issue config updates to the envoys based
+// on config seen on the configmap
+func (c *CRDClient) startMeshConfigListener(stop <-chan struct{}) {
+	// Create the subscription channel synchronously
+	cfgSubChannel := events.GetPubSubInstance().Subscribe(
+		announcements.MeshConfigAdded,
+		announcements.MeshConfigDeleted,
+		announcements.MeshConfigUpdated,
+	)
+
+	go func() {
+		// Defer unsubscription on async routine exit
+		defer events.GetPubSubInstance().Unsub(cfgSubChannel)
+
+		for {
+			select {
+			case msg := <-cfgSubChannel:
+				psubMsg, ok := msg.(events.PubSubMessage)
+				if !ok {
+					log.Error().Msgf("Could not cast pubsub message")
+					continue
+				}
+
+				switch psubMsg.AnnouncementType {
+				case announcements.MeshConfigAdded:
+					log.Debug().Msgf("[%s] OSM MeshConfig added event triggered a global proxy broadcast",
+						psubMsg.AnnouncementType)
+					events.GetPubSubInstance().Publish(events.PubSubMessage{
+						AnnouncementType: announcements.ScheduleProxyBroadcast,
+						OldObj:           nil,
+						NewObj:           nil,
+					})
+
+				case announcements.MeshConfigDeleted:
+					// Ignore deletion. We expect config to be present
+					log.Debug().Msgf("[%s] OSM MeshConfig deleted event triggered a global proxy broadcast",
+						psubMsg.AnnouncementType)
+					events.GetPubSubInstance().Publish(events.PubSubMessage{
+						AnnouncementType: announcements.ScheduleProxyBroadcast,
+						OldObj:           nil,
+						NewObj:           nil,
+					})
```

Comment:
I'd say deletion is ok.  With the current ConfigMap implementation - you can delete the ConfigMap and OSM will continue to function because we have default values for all keys.
This should be the case for the Config CRD as well.  If there is no OSM Config object present - we should rely on the default values for these (already defined for current Configmap - no reason not to reuse them).

Reply:
Summary of our offline discussion: My personal preference is to use an explicit CRD for the configuration, to avoid hiding default values in the code. Thus, one (or at least one) MeshConfig CRD object has to be presented in the mesh, if this is doable in k8s.